src/messageforwarder.py

```diff
@@ -0,0 +1,241 @@
+from collections import OrderedDict
+from threading import Lock
+
+
+class MessageForwarder:
+    """
+    Handles message forwarding between bouncer DMs and server staff.
+
+    There are two key parts of this functionality:
+     - Bouncer receives DM -> forward it to staff
+     - Staff replies to forwarded message -> reply to user in DMS
+
+    The first part is here. For the second part, the reply command handles figuring out who to message and sending the message.
+
+    We could move reply command functionality into this class, but I left it as is.
+    """
+    def __init__(self, create_threads: bool, mailbox_channel: int, ban_appeal_channel: int, home_server: int):
+        """
+        Creates a new message forwarder.
+
+        :param create_threads: Whether to forward messages into threads or not.
+                               If true, threads are created under either the mailbox/ban appeal channel as appropriate.
+                               If false, messages are sent directly to either the mailbox/ban appeal channel as appropriate.
+        :param mailbox_channel: The channel to forward regular DMs to.
+        :param ban_appeal_channel: The channel to forward ban appeal DMs to.
+        :param home_server: Bouncer's home server, used to make thread names nice.
+        """
+        # Configuration
+        self._create_threads = create_threads
+        self._mailbox_channel = mailbox_channel
+        self._ban_appeal_channel = ban_appeal_channel
+        self._home_server = home_server
+
+        # Maps user ids to reply thread ids - used when receiving a DM to know which thread to forward it to (if create_threads is true)
+        self._user_id_to_thread_id = LRUCache(lambda user_id: db.get_user_reply_thread_id(user_id), maxsize=50)
+
+        # Maps user reply thread ids to user ids - used when staff replies in a thread to know which user to send the reply to (this will work for existing threads even if create_threads is false)
+        self._thread_id_to_user_id = LRUCache(lambda thread_id: db.get_user_reply_thread_user_id(thread_id), maxsize=50)
+
+        # Both of the above use an LRU cache wrapper around accessing the DB so that every DM/reply does not trigger DB access
+        # The chosen cache size should be equal to the number of concurrent/active conversations we expect users to have with bouncer
+
+    async def on_dm(self, message: discord.Message):
+        """
+        On a DM, forward the message to staff.
+
+        :param message: The message that was sent.
+        """
+        # Ignore blocked users
+        if commands.bu.is_in_blocklist(message.author.id):
+            return
+
+        # Figure out what message to send, which channel to send in, etc based on which server the sender is in
+        if is_in_home_server(message.author):  # If the user is in the home server, treat it as a regular DM
+            reply_message = f"<@{message.author.id}>"
+            answering_machine = commands.reply_am
```

Comment:
Is my Python knowledge lacking? How is this not a scoping issue?

Reply:
Yeah, you actually don't need to predeclare these. The scope of these local variables is the rest of the function.

I think this behavior is defined here: https://docs.python.org/3/reference/executionmodel.html

Particularly:
> A block is a piece of Python program text that is executed as a unit. The following are blocks: a module, a function body, and a class definition.
> ...
> A scope defines the visibility of a name within a block. *If a local variable is defined in a block, its scope includes that block.*

Apparently Python calls bits of code indented to the same level 'suites': https://docs.python.org/3/reference/compound_stmts.html

> A compound statement consists of one or more ‘clauses.’ A clause consists of a header and a ‘suite.’ The clause headers of a particular compound statement are all at the same indentation level. ... A suite is a group of statements controlled by a clause. A suite can be one or more semicolon-separated simple statements on the same line as the header, following the header’s colon, *or it can be one or more indented statements on subsequent lines.*

And these do not affect scope.

I looked up all these docs just know, and this is all kinda hard to read, so I could be interpreting them wrong, but the code does work.